[Bug]: Intermittent CI Failure

### PyBaMM Version

develop

### Python Version

3.10, 3.12

### Describe the bug

There is an intermittent failure on CI resulting from https://github.com/pybamm-team/PyBaMM/pull/5112

See: 
https://github.com/pybamm-team/PyBaMM/actions/runs/16673618028/job/47195320616?pr=5129
https://github.com/pybamm-team/PyBaMM/actions/runs/16605096852/job/46974781286

There are more if you look.

I think this happens on line 101 of the new SciKit FEM 3D code.

@Rishab87 Can you please take a look at this? It looks like the inverse mass matrix can be a scalar in some non-deterministic case, which causes this? This seems strange to me

Tagging @rtimms @kratman 

### Steps to Reproduce

This failure appears to be non-deterministic 😬 

### Relevant log output

```shell
see the linked PR's
```

Btw while going through past CI failures also noticed few more tests failing non-deterministically:
```python
FAILED tests/unit/test_experiments/test_simulation_with_experiment.py::TestSimulationExperiment::test_run_experiment_cccv_solvers
FAILED tests/unit/test_experiments/test_simulation_with_experiment.py::TestSimulationExperiment::test_cycle_summary_variables - NotImplementedError: BinaryOperator not implemented for symbols of type <class 'pybamm.expression_tree.binary_operators.Division'> and <class 'pybamm.expression_tree.broadcasts.PrimaryBroadcast'>
FAILED tests/unit/test_experiments/test_simulation_with_experiment.py::TestSimulationExperiment::test_run_experiment_half_cell - NotImplementedError: BinaryOperator not implemented for symbols of type <class 'pybamm.expression_tree.binary_operators.Multiplication'> and <class 'pybamm.expression_tree.scalar.Scalar'>
FAILED tests/unit/test_experiments/test_simulation_with_experiment.py::TestSimulationExperiment::test_run_experiment_lead_acid - NotImplementedError: BinaryOperator not implemented for symbols of type <class 'pybamm.expression_tree.binary_operators.Division'> and <class 'pybamm.expression_tree.parameter.Parameter'>
```
Apart from these there were few more tests like serilisation failing non-deterministically.

I'm not sure if this is expected behavior or what the main issue is, but I can try investigating. I just wanted to let you know. My best guess is definitely race conditions due to the order of operations during simplification.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@
 
 ## Bug fixes
 
+- Fix non-deterministic ShapeError in 3D FEM gradient method ([#5143](https://github.com/pybamm-team/PyBaMM/pull/5143))
 - Fixes negative electrode boundary values for half-cell voltage contributions. ([#5139](https://github.com/pybamm-team/PyBaMM/pull/5139))
 - Makes `A_cc` L_z * L_y * number of layers ([#5138](https://github.com/pybamm-team/PyBaMM/pull/5138))
 - Fixes `TimeIntegral` expression node summation when dependent on an input parameter. ([#5119](https://github.com/pybamm-team/PyBaMM/pull/5119))

--- a/src/pybamm/spatial_methods/scikit_finite_element_3d.py
+++ b/src/pybamm/spatial_methods/scikit_finite_element_3d.py
@@ -82,6 +82,7 @@ class ScikitFiniteElement3D(pybamm.SpatialMethod):
         :class:`pybamm.Concatenation`
             The 3D gradient as concatenation of x, y, z or z, r, theta components
         """
+
         skfem = import_optional_dependency("skfem")
         domain = symbol.domain[0]
         mesh = self.mesh[domain]
@@ -97,6 +98,12 @@ class ScikitFiniteElement3D(pybamm.SpatialMethod):
 
         mass = skfem.asm(mass_form, mesh.basis)
         mass_inv = pybamm.Matrix(inv(csc_matrix(mass)))
+
+        if isinstance(discretised_symbol, pybamm.Scalar):  # pragma: no cover
+            zeros = pybamm.Vector(np.zeros((mesh.npts, 1)))
+            grad = pybamm.Concatenation(zeros, zeros, zeros, check_domain=False)
+            grad.copy_domains(symbol)
+            return grad
 
         grad_x = mass_inv @ (grad_x_matrix @ discretised_symbol)
         grad_y = mass_inv @ (grad_y_matrix @ discretised_symbol)

